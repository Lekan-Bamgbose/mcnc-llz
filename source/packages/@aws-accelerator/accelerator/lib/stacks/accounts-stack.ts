/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { NagSuppressions } from 'cdk-nag';
import { pascalCase } from 'change-case';
import { Construct } from 'constructs';
import * as path from 'path';

import {
  Account,
  EnablePolicyType,
  Policy,
  PolicyAttachment,
  PolicyType,
  PolicyTypeEnum,
} from '@aws-accelerator/constructs';

import { Logger } from '../logger';
import { AcceleratorStack, AcceleratorStackProps } from './accelerator-stack';
import { PrepareStack } from './prepare-stack';

export interface AccountsStackProps extends AcceleratorStackProps {
  readonly configDirPath: string;
}

let key: cdk.aws_kms.Key;
export class AccountsStack extends AcceleratorStack {
  constructor(scope: Construct, id: string, props: AccountsStackProps) {
    super(scope, id, props);

    Logger.debug(`[accounts-stack] Region: ${cdk.Stack.of(this).region}`);

    let globalRegion = 'us-east-1';
    if (props.partition === 'aws-us-gov') {
      globalRegion = 'us-gov-west-1';
    }
    if (props.partition === 'aws-cn') {
      globalRegion = 'cn-north-1';
    }

    // Use existing management account key if in the home region
    // otherwise create new kms key
    if (props.globalConfig.homeRegion == cdk.Stack.of(this).region) {
      const keyArn = cdk.aws_ssm.StringParameter.valueForStringParameter(
        this,
        PrepareStack.MANAGEMENT_KEY_ARN_PARAMETER_NAME,
      );
      key = cdk.aws_kms.Key.fromKeyArn(this, 'ManagementKey', keyArn) as cdk.aws_kms.Key;
    } else {
      key = new cdk.aws_kms.Key(this, 'ManagementKey', {
        alias: 'alias/accelerator/management/kms/key',
        description: 'AWS Accelerator Management Account Kms Key',
        enableKeyRotation: true,
        removalPolicy: cdk.RemovalPolicy.RETAIN,
      });

      // Allow Accelerator Role to use the encryption key
      key.addToResourcePolicy(
        new cdk.aws_iam.PolicyStatement({
          sid: `Allow Accelerator Role in this account to use the encryption key`,
          principals: [new cdk.aws_iam.AnyPrincipal()],
          actions: ['kms:Encrypt', 'kms:Decrypt', 'kms:ReEncrypt*', 'kms:GenerateDataKey*', 'kms:DescribeKey'],
          resources: ['*'],
          conditions: {
            ArnLike: {
              'aws:PrincipalARN': [
                `arn:${cdk.Stack.of(this).partition}:iam::${cdk.Stack.of(this).account}:role/AWSAccelerator-*`,
              ],
            },
          },
        }),
      );

      // Allow Cloudwatch logs to use the encryption key
      key.addToResourcePolicy(
        new cdk.aws_iam.PolicyStatement({
          sid: `Allow Cloudwatch logs to use the encryption key`,
          principals: [new cdk.aws_iam.ServicePrincipal(`logs.${cdk.Stack.of(this).region}.amazonaws.com`)],
          actions: ['kms:Encrypt*', 'kms:Decrypt*', 'kms:ReEncrypt*', 'kms:GenerateDataKey*', 'kms:Describe*'],
          resources: ['*'],
        }),
      );

      new cdk.aws_ssm.StringParameter(this, 'AcceleratorManagementKmsArnParameter', {
        parameterName: PrepareStack.MANAGEMENT_KEY_ARN_PARAMETER_NAME,
        stringValue: key.keyArn,
      });
    }

    //
    // Global Organizations actions
    //
    if (globalRegion === cdk.Stack.of(this).region && props.partition != 'aws-cn') {
      if (props.organizationConfig.enable) {
        const enablePolicyTypeScp = new EnablePolicyType(this, 'enablePolicyTypeScp', {
          policyType: PolicyTypeEnum.SERVICE_CONTROL_POLICY,
          kmsKey: key,
          logRetentionInDays: props.globalConfig.cloudwatchLogRetentionInDays,
        });

        // Invite Accounts to Organization (GovCloud)
        const accountMap: Map<string, Account> = new Map<string, Account>();
        for (const account of [...props.accountsConfig.mandatoryAccounts, ...props.accountsConfig.workloadAccounts]) {
          Logger.info(`[accounts-stack] Ensure ${account.name} is part of the Organization`);

          const organizationAccount = new Account(this, pascalCase(`${account.name}OrganizationAccount`), {
            accountId: props.accountsConfig.getAccountId(account.name),
            assumeRoleName: props.globalConfig.managementAccountAccessRole,
            kmsKey: key,
            logRetentionInDays: props.globalConfig.cloudwatchLogRetentionInDays,
          });

          accountMap.set(account.name, organizationAccount);
        }

        // Deploy SCPs
        let quarantineScpId = '';
        for (const serviceControlPolicy of props.organizationConfig.serviceControlPolicies) {
          Logger.info(`[accounts-stack] Adding service control policy (${serviceControlPolicy.name})`);

          const scp = new Policy(this, serviceControlPolicy.name, {
            description: serviceControlPolicy.description,
            name: serviceControlPolicy.name,
            path: path.join(props.configDirPath, serviceControlPolicy.policy),
            type: PolicyType.SERVICE_CONTROL_POLICY,
            kmsKey: key,
            logRetentionInDays: props.globalConfig.cloudwatchLogRetentionInDays,
            acceleratorPrefix: 'AWSAccelerator',
            managementAccountAccessRole: props.globalConfig.managementAccountAccessRole,
          });
          scp.node.addDependency(enablePolicyTypeScp);

          if (
            serviceControlPolicy.name == props.organizationConfig.quarantineNewAccounts?.scpPolicyName &&
            props.partition == 'aws'
          ) {
            new cdk.aws_ssm.StringParameter(this, pascalCase(`SsmParam${scp.name}ScpPolicyId`), {
              parameterName: `/accelerator/organizations/scp/${scp.name}/id`,
              stringValue: scp.id,
            });
            quarantineScpId = scp.id;
          }

          for (const organizationalUnit of serviceControlPolicy.deploymentTargets.organizationalUnits ?? []) {
            Logger.info(
              `[accounts-stack] Attaching service control policy (${serviceControlPolicy.name}) to organizational unit (${organizationalUnit})`,
            );

            new PolicyAttachment(this, pascalCase(`Attach_${scp.name}_${organizationalUnit}`), {
              policyId: scp.id,
              targetId: props.organizationConfig.getOrganizationalUnitId(organizationalUnit),
              type: PolicyType.SERVICE_CONTROL_POLICY,
              kmsKey: key,
              logRetentionInDays: props.globalConfig.cloudwatchLogRetentionInDays,
            });
          }

          for (const account of serviceControlPolicy.deploymentTargets.accounts ?? []) {
            const policyAttachment = new PolicyAttachment(this, pascalCase(`Attach_${scp.name}_${account}`), {
              policyId: scp.id,
              targetId: props.accountsConfig.getAccountId(account),
              type: PolicyType.SERVICE_CONTROL_POLICY,
              kmsKey: key,
              logRetentionInDays: props.globalConfig.cloudwatchLogRetentionInDays,
            });

            // Add dependency to ensure that account is part of the OU before
            // attempting to add the SCP
            const organizationAccount = accountMap.get(account);
            if (organizationAccount) {
              policyAttachment.node.addDependency(organizationAccount);
            }
          }
        }

        if (props.organizationConfig.quarantineNewAccounts?.enable === true && props.partition == 'aws') {
          // Create resources to attach quarantine scp to
          // new accounts created in organizations
          Logger.info(`[accounts-stack] Creating resources to quarantine new accounts`);
          const orgPolicyRead = new cdk.aws_iam.PolicyStatement({
            sid: 'OrgRead',
            effect: cdk.aws_iam.Effect.ALLOW,
            actions: ['organizations:ListPolicies', 'organizations:DescribeCreateAccountStatus'],
            resources: ['*'],
          });

          const orgPolicyWrite = new cdk.aws_iam.PolicyStatement({
            sid: 'OrgWrite',
            effect: cdk.aws_iam.Effect.ALLOW,
            actions: ['organizations:AttachPolicy'],
            resources: [
              `arn:${
                this.partition
              }:organizations::${props.accountsConfig.getManagementAccountId()}:policy/o-*/service_control_policy/${quarantineScpId}`,
              `arn:${this.partition}:organizations::${props.accountsConfig.getManagementAccountId()}:account/o-*/*`,
            ],
          });

          Logger.info(`[accounts-stack] Creating function to attach quarantine scp to accounts`);
          const attachQuarantineFunction = new cdk.aws_lambda.Function(this, 'AttachQuarantineScpFunction', {
            code: cdk.aws_lambda.Code.fromAsset(path.join(__dirname, '../lambdas/attach-quarantine-scp/dist')),
            runtime: cdk.aws_lambda.Runtime.NODEJS_14_X,
            handler: 'index.handler',
            description: 'Lambda function to attach quarantine scp to new accounts',
            timeout: cdk.Duration.minutes(5),
            environment: { SCP_POLICY_NAME: props.organizationConfig.quarantineNewAccounts?.scpPolicyName ?? '' },
            environmentEncryption: key,
            initialPolicy: [orgPolicyRead, orgPolicyWrite],
          });

          // AwsSolutions-IAM4: The IAM user, role, or group uses AWS managed policies
          NagSuppressions.addResourceSuppressionsByPath(
            this,
            `${this.stackName}/AttachQuarantineScpFunction/ServiceRole/Resource`,
            [
              {
                id: 'AwsSolutions-IAM4',
                reason: 'AWS Custom resource provider framework-role created by cdk.',
              },
            ],
          );

          // AwsSolutions-IAM5: The IAM entity contains wildcard permissions
          NagSuppressions.addResourceSuppressionsByPath(
            this,
            `${this.stackName}/AttachQuarantineScpFunction/ServiceRole/DefaultPolicy/Resource`,
            [
              {
                id: 'AwsSolutions-IAM5',
                reason: 'Allows only specific policy.',
              },
            ],
          );

          const createAccountEventRule = new cdk.aws_events.Rule(this, 'CreateAccountRule', {
            eventPattern: {
              source: ['aws.organizations'],
              detailType: ['AWS API Call via CloudTrail'],
              detail: {
                eventSource: ['organizations.amazonaws.com'],
                eventName: ['CreateAccount'],
              },
            },
            description: 'Rule to notify when a new account is created.',
          });

          createAccountEventRule.addTarget(
            new cdk.aws_events_targets.LambdaFunction(attachQuarantineFunction, {
              maxEventAge: cdk.Duration.hours(4),
              retryAttempts: 2,
            }),
          );

          //If any GovCloud accounts are configured also
          //watch for any GovCloudCreateAccount events
          if (props.accountsConfig.anyGovCloudAccounts()) {
            Logger.info(
              `[accounts-stack] Creating EventBridge rule to attach quarantine scp to accounts when GovCloud is enabled`,
            );
            const createGovCloudAccountEventRule = new cdk.aws_events.Rule(this, 'CreateGovCloudAccountRule', {
              eventPattern: {
                source: ['aws.organizations'],
                detailType: ['AWS API Call via CloudTrail'],
                detail: {
                  eventSource: ['organizations.amazonaws.com'],
                  eventName: ['CreateGovCloudAccount'],
                },
              },
              description: 'Rule to notify when a new account is created using the create govcloud account api.',
            });

            createGovCloudAccountEventRule.addTarget(
              new cdk.aws_events_targets.LambdaFunction(attachQuarantineFunction, {
                maxEventAge: cdk.Duration.hours(4),
                retryAttempts: 2,
              }),
            );
          }

          new cdk.aws_logs.LogGroup(this, `${attachQuarantineFunction.node.id}LogGroup`, {
            logGroupName: `/aws/lambda/${attachQuarantineFunction.functionName}`,
            retention: props.globalConfig.cloudwatchLogRetentionInDays,
            encryptionKey: key,
            removalPolicy: cdk.RemovalPolicy.DESTROY,
          });
        }
      }
    }
    Logger.info('[accounts-stack] Completed stack synthesis');
  }
}
