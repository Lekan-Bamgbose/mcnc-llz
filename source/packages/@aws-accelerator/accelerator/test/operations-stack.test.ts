/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { SynthUtils } from '@aws-cdk/assert';
import { OperationsStack } from '../lib/stacks/operations-stack';
import { AcceleratorStackNames } from '../lib/accelerator';
import { AcceleratorStage } from '../lib/accelerator-stage';
import {
  ACCOUNT_CONFIG,
  GLOBAL_CONFIG,
  IAM_CONFIG,
  NETWORK_CONFIG,
  ORGANIZATION_CONFIG,
  SECURITY_CONFIG,
} from './configs/test-config';
import * as path from 'path';
import { AcceleratorStackProps } from '../lib/stacks/accelerator-stack';

const testNamePrefix = 'Construct(OperationsStack): ';

/**
 * OperationsStack
 */
const app = new cdk.App({
  context: { 'config-dir': path.join(__dirname, 'configs') },
});
const configDirPath = app.node.tryGetContext('config-dir');

const env = {
  account: '333333333333',
  region: 'us-east-1',
};

const props: AcceleratorStackProps = {
  env,
  configDirPath,
  accountsConfig: ACCOUNT_CONFIG,
  globalConfig: GLOBAL_CONFIG,
  iamConfig: IAM_CONFIG,
  networkConfig: NETWORK_CONFIG,
  organizationConfig: ORGANIZATION_CONFIG,
  securityConfig: SECURITY_CONFIG,
  partition: 'aws',
};

const stack = new OperationsStack(
  app,
  `${AcceleratorStackNames[AcceleratorStage.OPERATIONS]}-${env.account}-${env.region}`,
  props,
);

/**
 * OperationsStack construct test
 */
describe('OperationsStack', () => {
  /**
   * Snapshot test
   */
  test(`${testNamePrefix} Snapshot Test`, () => {
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  });

  /**
   * Number of IAM group resource test
   */
  test(`${testNamePrefix} IAM group resource count test`, () => {
    cdk.assertions.Template.fromStack(stack).resourceCountIs('AWS::IAM::Group', 1);
  });

  /**
   * Number of IAM user resource test
   */
  test(`${testNamePrefix} IAM user resource count test`, () => {
    cdk.assertions.Template.fromStack(stack).resourceCountIs('AWS::IAM::User', 2);
  });

  /**
   * Number of SecretsManager secret resource test
   */
  test(`${testNamePrefix} SecretsManager secret resource count test`, () => {
    cdk.assertions.Template.fromStack(stack).resourceCountIs('AWS::SecretsManager::Secret', 2);
  });

  /**
   * Number of IAM managedPolicy resource test
   */
  test(`${testNamePrefix} IAM managedPolicy resource count test`, () => {
    cdk.assertions.Template.fromStack(stack).resourceCountIs('AWS::IAM::ManagedPolicy', 1);
  });

  /**
   * Number of IAM role resource test
   */
  test(`${testNamePrefix} IAM role resource count test`, () => {
    cdk.assertions.Template.fromStack(stack).resourceCountIs('AWS::IAM::Role', 1);
  });

  /**
   * Number of IAM InstanceProfile resource test
   */
  test(`${testNamePrefix} IAM InstanceProfile resource count test`, () => {
    cdk.assertions.Template.fromStack(stack).resourceCountIs('AWS::IAM::InstanceProfile', 1);
  });

  /**
   * Number of IAM SAMLProvider resource test
   */
  test(`${testNamePrefix} IAM SAMLProvider resource count test`, () => {
    cdk.assertions.Template.fromStack(stack).resourceCountIs('AWS::IAM::SAMLProvider', 1);
  });

  /**
   * Number of SSM parameter resource test
   */
  test(`${testNamePrefix} SSM parameter resource count test`, () => {
    cdk.assertions.Template.fromStack(stack).resourceCountIs('AWS::SSM::Parameter', 2);
  });

  /**
   * IAM group Administrators resource configuration test
   */
  test(`${testNamePrefix} IAM group Administrators resource configuration test`, () => {
    cdk.assertions.Template.fromStack(stack).templateMatches({
      Resources: {
        AdministratorsA37EF73A: {
          Type: 'AWS::IAM::Group',
          Properties: {
            GroupName: 'Administrators',
            ManagedPolicyArns: [
              {
                'Fn::Join': [
                  '',
                  [
                    'arn:',
                    {
                      Ref: 'AWS::Partition',
                    },
                    ':iam::aws:policy/AdministratorAccess',
                  ],
                ],
              },
            ],
          },
        },
      },
    });
  });

  /**
   * IAM user BreakGlassUser01 resource configuration test
   */
  test(`${testNamePrefix} IAM user BreakGlassUser01 resource configuration test`, () => {
    cdk.assertions.Template.fromStack(stack).templateMatches({
      Resources: {
        BreakGlassUser01AA051328: {
          Type: 'AWS::IAM::User',
          Properties: {
            Groups: [
              {
                Ref: 'AdministratorsA37EF73A',
              },
            ],
            LoginProfile: {
              Password: {
                'Fn::Join': [
                  '',
                  [
                    '{{resolve:secretsmanager:',
                    {
                      Ref: 'BreakGlassUser01Secret8A54324D',
                    },
                    ':SecretString:::}}',
                  ],
                ],
              },
            },
            PermissionsBoundary: {
              Ref: 'DefaultBoundaryPolicy489A8D26',
            },
            UserName: 'breakGlassUser01',
          },
        },
      },
    });
  });

  /**
   * SecretsManager secret BreakGlassUser01Secret resource configuration test
   */
  test(`${testNamePrefix} SecretsManager secret BreakGlassUser01Secret resource configuration test`, () => {
    cdk.assertions.Template.fromStack(stack).templateMatches({
      Resources: {
        BreakGlassUser01Secret8A54324D: {
          Type: 'AWS::SecretsManager::Secret',
          UpdateReplacePolicy: 'Delete',
          DeletionPolicy: 'Delete',
          Properties: {
            GenerateSecretString: {
              GenerateStringKey: 'password',
              SecretStringTemplate: '{"username":"breakGlassUser01"}',
            },
            Name: '/accelerator/breakGlassUser01',
          },
        },
      },
    });
  });

  /**
   * IAM user BreakGlassUser02 resource configuration test
   */
  test(`${testNamePrefix} IAM user BreakGlassUser02 resource configuration test`, () => {
    cdk.assertions.Template.fromStack(stack).templateMatches({
      Resources: {
        BreakGlassUser02DFF444C8: {
          Type: 'AWS::IAM::User',
          Properties: {
            Groups: [
              {
                Ref: 'AdministratorsA37EF73A',
              },
            ],
            LoginProfile: {
              Password: {
                'Fn::Join': [
                  '',
                  [
                    '{{resolve:secretsmanager:',
                    {
                      Ref: 'BreakGlassUser02Secret4D200D8D',
                    },
                    ':SecretString:::}}',
                  ],
                ],
              },
            },
            PermissionsBoundary: {
              Ref: 'DefaultBoundaryPolicy489A8D26',
            },
            UserName: 'breakGlassUser02',
          },
        },
      },
    });
  });

  /**
   * SecretsManager secret BreakGlassUser02Secret resource configuration test
   */
  test(`${testNamePrefix} SecretsManager secret BreakGlassUser02Secret resource configuration test`, () => {
    cdk.assertions.Template.fromStack(stack).templateMatches({
      Resources: {
        BreakGlassUser02Secret4D200D8D: {
          Type: 'AWS::SecretsManager::Secret',
          UpdateReplacePolicy: 'Delete',
          DeletionPolicy: 'Delete',
          Properties: {
            GenerateSecretString: {
              GenerateStringKey: 'password',
              SecretStringTemplate: '{"username":"breakGlassUser02"}',
            },
            Name: '/accelerator/breakGlassUser02',
          },
        },
      },
    });
  });

  /**
   * IAM managedPolicy DefaultBoundaryPolicy resource configuration test
   */
  test(`${testNamePrefix} IAM managedPolicy DefaultBoundaryPolicy resource configuration test`, () => {
    cdk.assertions.Template.fromStack(stack).templateMatches({
      Resources: {
        DefaultBoundaryPolicy489A8D26: {
          Type: 'AWS::IAM::ManagedPolicy',
          Properties: {
            Description: '',
            ManagedPolicyName: 'Default-Boundary-Policy',
            Path: '/',
            PolicyDocument: {
              Statement: [
                {
                  Action: '*',
                  Effect: 'Allow',
                  Resource: '*',
                },
              ],
              Version: '2012-10-17',
            },
          },
        },
      },
    });
  });

  /**
   * IAM role Ec2DefaultSsmAdRole resource configuration test
   */
  test(`${testNamePrefix} IAM role Ec2DefaultSsmAdRole resource configuration test`, () => {
    cdk.assertions.Template.fromStack(stack).templateMatches({
      Resources: {
        Ec2DefaultSsmAdRoleADFFA4C6: {
          Type: 'AWS::IAM::Role',
          Properties: {
            AssumeRolePolicyDocument: {
              Statement: [
                {
                  Action: 'sts:AssumeRole',
                  Effect: 'Allow',
                  Principal: {
                    Service: 'ec2.amazonaws.com.cn',
                  },
                },
              ],
              Version: '2012-10-17',
            },
            ManagedPolicyArns: [
              {
                'Fn::Join': [
                  '',
                  [
                    'arn:',
                    {
                      Ref: 'AWS::Partition',
                    },
                    ':iam::aws:policy/AmazonSSMManagedInstanceCore',
                  ],
                ],
              },
              {
                'Fn::Join': [
                  '',
                  [
                    'arn:',
                    {
                      Ref: 'AWS::Partition',
                    },
                    ':iam::aws:policy/AmazonSSMDirectoryServiceAccess',
                  ],
                ],
              },
              {
                'Fn::Join': [
                  '',
                  [
                    'arn:',
                    {
                      Ref: 'AWS::Partition',
                    },
                    ':iam::aws:policy/CloudWatchAgentServerPolicy',
                  ],
                ],
              },
            ],
            PermissionsBoundary: {
              Ref: 'DefaultBoundaryPolicy489A8D26',
            },
            RoleName: 'EC2-Default-SSM-AD-Role',
          },
        },
      },
    });
  });

  /**
   * IAM InstanceProfile Ec2DefaultSsmAdRoleInstanceProfile resource configuration test
   */
  test(`${testNamePrefix} IAM InstanceProfile Ec2DefaultSsmAdRoleInstanceProfile resource configuration test`, () => {
    cdk.assertions.Template.fromStack(stack).templateMatches({
      Resources: {
        Ec2DefaultSsmAdRoleInstanceProfile: {
          Type: 'AWS::IAM::InstanceProfile',
          Properties: {
            InstanceProfileName: {
              Ref: 'Ec2DefaultSsmAdRoleADFFA4C6',
            },
            Roles: [
              {
                Ref: 'Ec2DefaultSsmAdRoleADFFA4C6',
              },
            ],
          },
        },
      },
    });
  });

  /**
   * IAM SAMLProvider ProviderSamlProvider resource configuration test
   */
  test(`${testNamePrefix} IAM SAMLProvider ProviderSamlProvider resource configuration test`, () => {
    cdk.assertions.Template.fromStack(stack).templateMatches({
      Resources: {
        ProviderSamlProviderDA84AD16: {
          Type: 'AWS::IAM::SAMLProvider',
          Properties: {
            Name: 'provider',
            SamlMetadataDocument: '',
          },
        },
      },
    });
  });

  /**
   * SSM parameter SsmParamStackId resource configuration test
   */
  test(`${testNamePrefix} SSM parameter SsmParamStackId resource configuration test`, () => {
    cdk.assertions.Template.fromStack(stack).templateMatches({
      Resources: {
        SsmParamStackId521A78D3: {
          Type: 'AWS::SSM::Parameter',
          Properties: {
            Name: '/accelerator/AWSAccelerator-OperationsStack-333333333333-us-east-1/stack-id',
            Type: 'String',
            Value: {
              Ref: 'AWS::StackId',
            },
          },
        },
      },
    });
  });
});
