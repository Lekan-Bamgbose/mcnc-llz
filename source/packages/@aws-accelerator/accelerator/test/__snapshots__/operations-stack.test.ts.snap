// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OperationsStack Construct(OperationsStack):  Snapshot Test 1`] = `
Object {
  "Resources": Object {
    "AdministratorsA37EF73A": Object {
      "Metadata": Object {
        "cdk_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "AwsSolutions-IAM4",
              "reason": "Groups created as per accelerator iam-config needs AWS managed policy",
            },
          ],
        },
      },
      "Properties": Object {
        "GroupName": "Administrators",
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AdministratorAccess",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Group",
    },
    "BreakGlassUser01AA051328": Object {
      "Properties": Object {
        "Groups": Array [
          Object {
            "Ref": "AdministratorsA37EF73A",
          },
        ],
        "LoginProfile": Object {
          "Password": Object {
            "Fn::Join": Array [
              "",
              Array [
                "{{resolve:secretsmanager:",
                Object {
                  "Ref": "BreakGlassUser01Secret8A54324D",
                },
                ":SecretString:::}}",
              ],
            ],
          },
        },
        "PermissionsBoundary": Object {
          "Ref": "DefaultBoundaryPolicy489A8D26",
        },
        "UserName": "breakGlassUser01",
      },
      "Type": "AWS::IAM::User",
    },
    "BreakGlassUser01Secret8A54324D": Object {
      "DeletionPolicy": "Delete",
      "Metadata": Object {
        "cdk_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "AwsSolutions-SMG4",
              "reason": "Accelerator users created as per iam-config file, upcoming change will take care of secret automatic rotation",
            },
          ],
        },
      },
      "Properties": Object {
        "GenerateSecretString": Object {
          "GenerateStringKey": "password",
          "SecretStringTemplate": "{\\"username\\":\\"breakGlassUser01\\"}",
        },
        "Name": "/accelerator/breakGlassUser01",
      },
      "Type": "AWS::SecretsManager::Secret",
      "UpdateReplacePolicy": "Delete",
    },
    "BreakGlassUser02DFF444C8": Object {
      "Properties": Object {
        "Groups": Array [
          Object {
            "Ref": "AdministratorsA37EF73A",
          },
        ],
        "LoginProfile": Object {
          "Password": Object {
            "Fn::Join": Array [
              "",
              Array [
                "{{resolve:secretsmanager:",
                Object {
                  "Ref": "BreakGlassUser02Secret4D200D8D",
                },
                ":SecretString:::}}",
              ],
            ],
          },
        },
        "PermissionsBoundary": Object {
          "Ref": "DefaultBoundaryPolicy489A8D26",
        },
        "UserName": "breakGlassUser02",
      },
      "Type": "AWS::IAM::User",
    },
    "BreakGlassUser02Secret4D200D8D": Object {
      "DeletionPolicy": "Delete",
      "Metadata": Object {
        "cdk_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "AwsSolutions-SMG4",
              "reason": "Accelerator users created as per iam-config file, upcoming change will take care of secret automatic rotation",
            },
          ],
        },
      },
      "Properties": Object {
        "GenerateSecretString": Object {
          "GenerateStringKey": "password",
          "SecretStringTemplate": "{\\"username\\":\\"breakGlassUser02\\"}",
        },
        "Name": "/accelerator/breakGlassUser02",
      },
      "Type": "AWS::SecretsManager::Secret",
      "UpdateReplacePolicy": "Delete",
    },
    "DefaultBoundaryPolicy489A8D26": Object {
      "Metadata": Object {
        "cdk_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "AwsSolutions-IAM5",
              "reason": "Policies definition are derived from accelerator iam-config boundary-policy file",
            },
          ],
        },
      },
      "Properties": Object {
        "Description": "",
        "ManagedPolicyName": "Default-Boundary-Policy",
        "Path": "/",
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "*",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::ManagedPolicy",
    },
    "Ec2DefaultSsmAdRoleADFFA4C6": Object {
      "Metadata": Object {
        "cdk_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "AwsSolutions-IAM4",
              "reason": "IAM Role created as per accelerator iam-config needs AWS managed policy",
            },
          ],
        },
      },
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "ec2.amazonaws.com.cn",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AmazonSSMManagedInstanceCore",
              ],
            ],
          },
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AmazonSSMDirectoryServiceAccess",
              ],
            ],
          },
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/CloudWatchAgentServerPolicy",
              ],
            ],
          },
        ],
        "PermissionsBoundary": Object {
          "Ref": "DefaultBoundaryPolicy489A8D26",
        },
        "RoleName": "EC2-Default-SSM-AD-Role",
      },
      "Type": "AWS::IAM::Role",
    },
    "Ec2DefaultSsmAdRoleInstanceProfile": Object {
      "Properties": Object {
        "InstanceProfileName": Object {
          "Ref": "Ec2DefaultSsmAdRoleADFFA4C6",
        },
        "Roles": Array [
          Object {
            "Ref": "Ec2DefaultSsmAdRoleADFFA4C6",
          },
        ],
      },
      "Type": "AWS::IAM::InstanceProfile",
    },
    "ProviderSamlProviderDA84AD16": Object {
      "Properties": Object {
        "Name": "provider",
        "SamlMetadataDocument": "",
      },
      "Type": "AWS::IAM::SAMLProvider",
    },
    "SsmParamAcceleratorVersionFF83282D": Object {
      "Properties": Object {
        "Name": "/accelerator/AWSAccelerator-OperationsStack-333333333333-us-east-1/version",
        "Type": "String",
        "Value": "1.0.1",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "SsmParamStackId521A78D3": Object {
      "Properties": Object {
        "Name": "/accelerator/AWSAccelerator-OperationsStack-333333333333-us-east-1/stack-id",
        "Type": "String",
        "Value": Object {
          "Ref": "AWS::StackId",
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
  },
}
`;
